### ISSUE TITLE
Validation fails if attribute is persisted to post content & meta.

### ISSUE THREAD

**Issue by Shelob9** [WordPress/gutenberg#4989]
If a block saves markup with a representation of an attribute in HTML and that same attribute as meta, block validation fails. The meta is saved, the markup is good in post content, block validation fails.

## Issue Overview
Consider this requirement for a block:
1) Must save completely valid HTML markup to post content that, when the client has no JavaScript, is totally human and machine (assistive devices, search engines, etc) readable, and can be progressively enhanced by front-end JavaScript/CSS.
2) The fields represented in that HTML markup  should be queryable in the database.

Requirement 1 is provided by a Gutenberg static HTML block and requirement 2 is satisfied by the WordPress Meta API if the field is saved as post meta.

In my mind, saving an attribute as post meta, and static HTML representing its value in a semantic, human readable form is best for front-end requirements -- accessibility, no JS fallbacks, SEO, the philosophy of progressive enhancement -- and back-end needs -- we need to query by field data and update field data.

BTW I'm aware that this duplicates data. Given that we have a hook that runs after post meta is updated and we have a function for converting attributes to static HTML, that should be solvable.

## Steps to Reproduce (for bugs)
1. [Register a block with a meta attribute and register the meta field](https://gist.github.com/Shelob9/d0cdd75ee6479691e92f993ee1470842#file-block-php)
2. [Create block edit function  so that you can edit attribute and see that setAttributes is updating state correctly](https://gist.github.com/Shelob9/d0cdd75ee6479691e92f993ee1470842#file-block-js-L9-L32). 
3.[Create block save function that for a non-meta attribute ](https://gist.github.com/Shelob9/d0cdd75ee6479691e92f993ee1470842#file-block-js-L33-L38). 

## Expected Behavior
No validation error.

## Current Behavior
Block validation fails.

## Possible Solution
I think the problem here is it's being parsed as if it's a an attribute being stored in markup/normal serialization, but I'm not sure. If so, can that be disabled in these situations? 

## Screenshots / Video
![screen shot 2018-02-10 at 2 07 12 pm](https://user-images.githubusercontent.com/1994311/36065705-8c2a905e-0e6c-11e8-8517-4999c7e77a63.png)
![screen shot 2018-02-10 at 2 07 30 pm](https://user-images.githubusercontent.com/1994311/36065706-8f3fa2c0-0e6c-11e8-9ab8-7f49e5fe8720.png)
## Todos
- [ ] Tests
- [ ] Documentation

**Comment by Shelob9** [comment]
I created a workaround, but it involves a second "fake" content attribute, which is sub-optimal: https://gist.github.com/Shelob9/36d7096d919995ae615d4d4bbf10ffbc

**Comment by youknowriad** [comment]
The problem is that the validation is executed while parsing the blocks, and the parsing don't know about the post object as it's supposed to be generic enough to be used in context with post information.

We might have to rethink adding specific block attributes sources like "meta" in a generic way where we could hook into the parser, the selector, the onChange etc... and enhance how these work.

cc @aduth

**Comment by aduth** [comment]
>We might have to rethink adding specific block attributes sources like "meta" in a generic way where we could hook into the parser, the selector, the onChange etc... and enhance how these work.

Strange coincidence, I was iterating on an idea here last night.

Approach I'd thought was a filter in `getBlockAttribute` which `editPost` adds a hook for and allows custom `source: "meta"` handling.

**Comment by aduth** [comment]
One related issue I'd come across in starting this is that many of our post selectors are in `editor`, which ought to be migrated (especially as we plan to expose all of them).

**Comment by designsimply** [comment]
Noting the console warning and error in plain text for reference:

```
Block validation: Expected token of type `EndTag`
Block validation: Block validation failed for `shelob9/metatest`
```

**Comment by aduth** [comment]
While this is still acknowledged to be a bug, it's not one I'd considered a blocker for 5.0, given both the workaround and considering that the circumstances under which it would impact the vast majority of users approaches near impossible, if one considers that it would become plainly obvious to be broken to a block implementer in the course of minimal testing.

#5077 can serve as incomplete prior art toward a possible resolution, and may in-fact tie to some potential optimizations, given that the meta attribute computation in the `getBlock` selector ([eliminated in #5077](https://github.com/WordPress/gutenberg/pull/5077/files#diff-4e1dd4776f75eb7a66b28280a56d8e0aL425)) has shown in recent performance audits to be a contributing factor to degraded experience.

**Comment by cbrakhage** [comment]
@aduth What is the work around to this? I'm a bit new to this, and I can't seem to figure out why my attribute isn't persisting after a page reload. It seems like initially my attribute isn't loaded, but when I delete the block and create a new one, both edit and save have the correct value.

**Comment by aduth** [comment]
@cbrakhage Have you looked over the code @Shelob9 had shared at https://github.com/WordPress/gutenberg/issues/4989#issuecomment-364682609 ? It seems to involve duplicating the attribute value as saved in both the markup of a block and the post meta.

**Comment by Horttcore** [comment]
When will this be fixed? I mean this is some serious bug and makes meta attributes useless

**Comment by epiqueras** [comment]
It looks like #16402 fixed this. I can't reproduce anymore.

**Comment by aduth** [comment]
@epiqueras If you're using the [sample plugin](https://gist.github.com/pento/19b35d621709042fc899e394a9387a54) I shared in #16402 as a base for testing, it doesn't include generated save output, so would not trigger the validation error observed here. In the sample, [changing this line](https://gist.github.com/pento/19b35d621709042fc899e394a9387a54#file-stars-block-js-L53) to `return el( Stars, { stars: arguments[ 0 ].attributes.stars } );` should prompt the issue.

In order to resolve this issue, there are still changes necessary after #16402 to separate validation from the initial parse to account for the applied meta attributes.

**Comment by epiqueras** [comment]
Strange, I did change it to `return props.attributes.stars`.

**Comment by aduth** [comment]
A fix is now proposed at #16569

**Comment by youknowriad** [comment]
As of #17153, meta attribute values are no longer initialized into parsed blocks, which invalidates this approach. The decision here is that meta attributes are not meant to be used on save functions since save functions are supposed to be pure and the meta value can change dynamically outside of the block.
